nio/client/async_client.py

```diff
@@ -903,25 +906,119 @@ async def login_raw(
+
+    async def register_with_token(
+        self,
+        username: str,
+        password: str,
+        registration_token: str,
+        device_name: str = "",
+    ) -> Union[RegisterResponse, RegisterErrorResponse]:
+        """Registers a user using a registration token.
+        See https://spec.matrix.org/latest/client-server-api/#token-authenticated-registration
+
+        Returns either a `RegisterResponse` if the request was successful or
+        a `RegisterErrorResponse` if there was an error with the request.
+
+        """
+        # must first register without token to get a session token
+        resp = await self.register_interactive(
+            username,
+            password,
+            auth_dict={"initial_device_display_name": self.device_id or "matrix-nio"},
+        )
+        if isinstance(resp, RegisterInteractiveError):
+            return RegisterErrorResponse(
+                resp.message, resp.status_code, resp.retry_after_ms, resp.soft_logout
+            )
+
+        # use session token to register with token
+        session_token = resp.session
+        resp = await self.register_interactive(
+            username,
+            password,
+            auth_dict={
+                "type": "m.login.registration_token",
+                "token": registration_token,
+                "session": session_token,
+            },
+        )
+        if isinstance(resp, RegisterInteractiveError):
+            return RegisterErrorResponse(
+                resp.message, resp.status_code, resp.retry_after_ms, resp.soft_logout
+            )
+
+        # finally call register with dummy auth with original session token
+        # to complete registration and acquire access token
+        return await self.register(
+            username, password, device_name=device_name, session_token=session_token
+        )
+
+    async def register(
+        self,
+        username: str,
+        password: str,
+        device_name: str = "",
+        session_token: Optional[str] = None,
+    ) -> Union[RegisterResponse, RegisterErrorResponse]:
+        """Register with homeserver.
+
+        Calls receive_response() to update the client state if necessary.
+
+        Args:
+            username (str): Username to register the new user as.
+            password (str): New password for the user.
+            device_name (str, optional): A display name to assign to a
+                newly-created device. Ignored if the logged in device
+                corresponds to a known device.
+            session_token (str, optional): The session token the server
+                provided during interactive registration. If not provided,
+                the session token is not added to the request's auth dict.
+
         Returns a 'RegisterResponse' if successful.
         """
+        auth_dict = {"type": "m.login.dummy"}
+        if session_token is not None:
+            auth_dict["session"] = session_token
```

Comment:
I'll need to read more on the flow, but it looked a little weird to me that with this, `auth_dict` could look like:
```
{"type": "m.login.dummy", "session": "some_token"}
```

Reply:
It is kinda weird, isn't it? It's really only necessary in an interactive approach [per the spec](https://spec.matrix.org/latest/client-server-api/#user-interactive-api-in-the-rest-api). I wanted to minimize the amount of code `register_with_token` required. The last thing that `register_with_token` does is use the `m.login.dummy` type which the `register` function already supported. I figured instead of duplicating `register`'s code, I could use it but have `register` optionally tack on the `session` id if provided.